enhancements/console/pinned-resources.md

```diff
@@ -0,0 +1,179 @@
+status: implementable
+---
+
+# Define pre-pinned resources
+
+## Summary
+
+The Developer perspective on the OpenShift Console provides pinned resources on the navigation which enables the user to go to the resource’s list page with a single click.
+
+Some customers want to provide defaults for those pre-pinned resources for all users.
+
+On the Developer perspective navigation, users can see the default pinned resources if they have not customized these pinned resources. The customers have no way to set these default pinned resources for their users.
+
+This enhancement allows cluster admins to define pre-pinned resources for new users and users who have not customized their navigation items.
+
+![Default pinned resources](default-pinned-resources.png)
+
+## Motivation
+
+Some customers want to provide default pre-pinned resources for all users.
+
+### User Stories
+
+#### Story 1
+
+Customer wants to customize the developer perspective navigation by providing pre-pinned resources to users who do not have their navigation items customized.
+
+As a cluster admin, I want to provide default pre-pinned resources for all users.
+
+### Goals
+
+- Allow cluster admins to define default pre-pinned resources on the dev perspective navigation
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+We can extend the existing `operator.openshift.io/v1` / `Console CRD`.
+We can add `spec.customization.perspectives.pinnedResources` which lists default pre-pinned resources
+
+### Workflow Description
+
+To add default pre-pinned resources:
+
+```text
+├── spec
+│   ├── customization
+|       ├── perspectives
+|           ├── id
+|           ├── ...
+|           ├── pinnedResources
+|               └── group
+                └── version
+                └── kind
+└── ...
+```
+
+Each perspective conforms to the following schema :
+
+- `pinnedResources: authV1.GroupVersionKind[]` , defines a list of default pinned resources
+- `group: string` , defines the group of the resource type.
+- `version: string` , defines the version of the resource type.
+- `kind: string` , defines the kind of the resource type.
+
+If the `pinnedResources` list is empty, the default pinned resources will be shown.
+If the `pinnedResources` list is non-empty, only the listed items will be shown if the user hasn't already customized the navigation items.
+
+### API Extensions
+
+Extend the existing `operator.openshift.io/v1` / `Console CRD` by adding `spec.customization.perspectives.pinnedResources`
+
+So the example yaml should look like this:
+
+To set default pre-pinned resources
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+  ...
+spec:
+  customization:
+    perspectives:
+    - id: 'dev'
+      ...
+      pinnedResources:
+      - group: apps
+        version: v1
+        kind: Deployment
+      - group: core
+        version: v1
+        kind: ConfigMap
+    - ...
+  ...
+```
+
+### Risks and Mitigations
+
+N/A
```

Comment:
Risk: A user provides an invalid combination of API/Group/Version
Risk: After a cluster is upgraded, a previously valid combination of API/Group/Version is now invalid

Reply:
@debsmita1 @JoelSpeed 

**Risk 1:** Invalid resources are ignored already today because of:
**Risk 2:** This could already happen today. The user can add a pinned resource and then delete a CRD or uninstall an operator. These resources are then not shown anymore in the sidebar.

But it makes sense to address both risks and mitigations here.

I see one more risk:

**Risk 3:** Currently the user preferences save the default pinned resources also if the user doesn't touch them. This means that the user settings always contain the info "Secret", "ConfigMap".

We should a) check if it's easy to fix that the defaults are not saved when the user doesn't touch them.

And b) Should we also use the cluster pinned resources when the pinned resources is (exactly) equal "Secret", "ConfigMap"?